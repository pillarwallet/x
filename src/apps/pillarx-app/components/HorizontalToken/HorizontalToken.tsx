import { useEffect, useState } from 'react';

// types
import { TokenContract } from '../../../../types/api';

// reducer
import { useGetBlockchainsListQuery } from '../../../token-atlas/api/token';

// components
import RandomAvatar from '../RandomAvatar/RandomAvatar';
import TokensPercentage from '../TokensPercentage/TokensPercentage';
import Body from '../Typography/Body';
import BodySmall from '../Typography/BodySmall';

type HorizontalTokenProps = {
  tokenIndex: number;
  tokenLogo?: string;
  tokenSymbol?: string;
  tokenName?: string;
  tokenValue?: number;
  tokenChains?: TokenContract[];
  percentage?: number;
  isLast?: boolean;
  onClick?: () => void;
};

const HorizontalToken = ({
  tokenIndex,
  tokenLogo,
  tokenSymbol,
  tokenName,
  tokenValue,
  tokenChains,
  percentage,
  isLast,
  onClick,
}: HorizontalTokenProps) => {
  const [blockchainLogo, setBlockchainLogo] = useState<string | undefined>(
    undefined
  );
  const [isBrokenImage, setIsBrokenImage] = useState<boolean>(false);

  const { data: blockchainListData, isSuccess } = useGetBlockchainsListQuery();

  useEffect(() => {
    if (tokenChains?.length === 1 && blockchainListData && isSuccess) {
      const chainData = blockchainListData?.result?.data.find(
        (chain) => chain.name === tokenChains[0].blockchain
      );
      setBlockchainLogo(chainData?.logo);
    }
  }, [blockchainListData, isSuccess, tokenChains]);

  return (
    <div
      className={`flex w-full justify-between py-5 cursor-pointer ${!isLast && 'border-b border-[#1F1D23] mobile:border-[#27262F]'} ${isLast && tokenIndex === 3 && 'mobile:border-b mobile:border-[#27262F]'}`}
      onClick={onClick}
    >
      <div className="flex items-center">
        <BodySmall className="text-purple_light mr-[18px]">
          0{tokenIndex}
        </BodySmall>
        <div className="relative w-[50px] h-[50px] rounded-full mr-3.5">
          {tokenLogo && !isBrokenImage ? (
            <img
              src={tokenLogo}
              alt="token-logo"
              className="w-full h-full object-fill rounded-full"
              data-testid="horizontal-token-logo"
              onError={() => setIsBrokenImage(true)}
            />
          ) : (
            <div className="w-full h-full overflow-hidden rounded-full">
              <RandomAvatar name={tokenName || ''} />
            </div>
          )}

          {/* Overlay text when no token logo available */}
          {(!tokenLogo || isBrokenImage) && (
            <span className="absolute inset-0 flex items-center justify-center text-white text-lg font-bold">
              {tokenName?.slice(0, 2)}
            </span>
          )}

          {/* Blockchain logo overlapping when only one blockchain for this token */}
          {tokenChains?.length === 1 ? (
            <div className="absolute bottom-0 right-0 w-6 h-6 rounded-full overflow-hidden border-[1px] bg-white border-container_grey transform translate-x-1/3 translate-y-1/3">
              <img
                src={blockchainLogo}
                alt="logo"
                className="w-full h-full object-contain"
              />
            </div>
          ) : null}
        </div>

        <div className="flex flex-col">
          {tokenSymbol && <Body>{tokenSymbol}</Body>}
          {tokenName && (
            <BodySmall className="text-purple_light">{tokenName}</BodySmall>
          )}
        </div>
      </div>
      <div className="flex flex-col items-center">
        {tokenValue && <Body>${tokenValue.toFixed(4)}</Body>}
        <TokensPercentage percentage={percentage} />
      </div>
    </div>
  );
};

export default HorizontalToken;
