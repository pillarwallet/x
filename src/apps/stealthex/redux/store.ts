import { configureStore, TypedStartListening } from '@reduxjs/toolkit';
import exchangeReducer from './reducers/exchange';
import { exchange } from '../lib/backend/api';

export const store = configureStore({
    reducer: {
        exchange: exchangeReducer,
        [exchange.reducerPath]: exchange.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware({ serializableCheck: false }).concat(exchange.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStartListening = TypedStartListening<RootState, AppDispatch>; 