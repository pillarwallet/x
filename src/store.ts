/* eslint-disable @typescript-eslint/no-use-before-define */
// Core
import {
  Middleware,
  Reducer,
  combineReducers,
  configureStore,
  createDynamicMiddleware,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

// Services
import swapSlice from './apps/the-exchange/reducer/theExchangeSlice';
import tokenAtlasSlice from './apps/token-atlas/reducer/tokenAtlasSlice';
import { pillarXApiPresence } from './services/pillarXApiPresence';
import { pillarXApiTransactionsHistory } from './services/pillarXApiTransactionsHistory';
import { pillarXApiWaitlist } from './services/pillarXApiWaitlist';

// Initialisation
const dynamicMiddleware = createDynamicMiddleware();
const middlewareReducers: { [key: string]: Reducer } = {};

/**
 * @name addReducer
 * @description addReducer is a function that allows anyone
 * working with PillarX to add reducers to the store
 * from their own application.
 *
 * @param newReducer
 */
export const addReducer = (newReducer: {
  reducerPath: string;
  reducer: Reducer;
}) => {
  middlewareReducers[newReducer.reducerPath as string] = newReducer.reducer;
  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * @name addMiddleware
 * @description addMiddleware is a function that allows anyone
 * working with PillarX to add RTK middleware to the store
 * from their own application.
 *
 * @param newMiddleware
 */
export const addMiddleware = (newMiddleware: {
  reducerPath: string;
  reducer: Reducer;
  middleware: Middleware;
}) => {
  middlewareReducers[newMiddleware.reducerPath as string] =
    newMiddleware.reducer;
  dynamicMiddleware.addMiddleware(newMiddleware.middleware);

  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * Export a store from RTK
 */
export const store = configureStore({
  // Empty reducer for now - the addMiddleware function
  // below will dynamically regenerate the reducers required
  // from the middleware functions.
  reducer: {},
  // Adding the api middleware enables caching, invalidation, polling,
  // and other useful features of `rtk-query`.
  // Note: here we have added dynamicMiddleware.middleware
  // which instructs the store to use the middleware that we
  // will add later.
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().prepend(dynamicMiddleware.middleware),
});

/**
 * Whilst we're here we'll immediately mount the
 * PillarXApi middleware. This will contain all the
 * core API functionality for the PillarX application.
 */
addMiddleware(pillarXApiWaitlist);
addMiddleware(pillarXApiPresence);
addMiddleware(pillarXApiTransactionsHistory);
addReducer(swapSlice);
addReducer(tokenAtlasSlice);

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

// // Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
