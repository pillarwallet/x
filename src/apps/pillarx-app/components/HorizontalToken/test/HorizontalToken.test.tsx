import { render, screen } from '@testing-library/react';
import renderer, { ReactTestRendererJSON } from 'react-test-renderer';

// provider
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// components
import HorizontalToken from '../HorizontalToken';

// Mock components
jest.mock('../../Typography/BodySmall', () => 'BodySmall');
jest.mock('../../Typography/Body', () => 'Body');
jest.mock('../../TokensPercentage/TokensPercentage', () => 'TokensPercentage');

describe('<HorizontalToken />', () => {
  const tokenIndex = 1;
  const logo = 'https://example.com/logo.png';
  const tokenSymbol = 'TNS';
  const tokenName = 'Token Name';
  const tokenValue = 123.4567;
  const percentage = 12.34;
  const isLast = false;

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <HorizontalToken
            tokenIndex={tokenIndex}
            tokenLogo={logo}
            tokenSymbol={tokenSymbol}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
            isLast={isLast}
          />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders the logo', () => {
    render(
      <Provider store={store}>
        <HorizontalToken
          tokenIndex={tokenIndex}
          tokenLogo={logo}
          tokenSymbol={tokenSymbol}
          tokenName={tokenName}
          tokenValue={tokenValue}
          percentage={percentage}
          isLast={isLast}
        />
      </Provider>
    );

    expect(screen.getByTestId('horizontal-token-logo')).toBeInTheDocument();
    expect(screen.queryByTestId('random-avatar')).not.toBeInTheDocument();
  });

  it('renders the default logo (random avatar) when logo is not provided', () => {
    render(
      <Provider store={store}>
        <HorizontalToken
          tokenIndex={tokenIndex}
          tokenSymbol={tokenSymbol}
          tokenName={tokenName}
          tokenValue={tokenValue}
          percentage={percentage}
          isLast={isLast}
        />
      </Provider>
    );

    expect(
      screen.queryByTestId('horizontal-token-logo')
    ).not.toBeInTheDocument();
    expect(screen.getByTestId('random-avatar')).toBeInTheDocument();
  });

  it('renders the name correctly', () => {
    render(
      <Provider store={store}>
        <HorizontalToken
          tokenIndex={tokenIndex}
          tokenLogo={logo}
          tokenSymbol={tokenSymbol}
          tokenName={tokenName}
          tokenValue={tokenValue}
          percentage={percentage}
          isLast={isLast}
        />
      </Provider>
    );

    expect(screen.getByText('Token Name')).toBeInTheDocument();
  });

  it('renders the token value correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <HorizontalToken
            tokenIndex={tokenIndex}
            tokenLogo={logo}
            tokenSymbol={tokenSymbol}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
            isLast={isLast}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;

    const divElement = tree.children?.filter(
      (child) => typeof child === 'object' && child.type === 'div'
    ) as ReactTestRendererJSON[];

    const tokenValueProp = divElement[1].children?.find(
      (child) => typeof child === 'object' && child.type === 'Body'
    ) as ReactTestRendererJSON;

    expect(tokenValueProp).not.toBeNull();
    expect(tokenValueProp.children).toContain(`${tokenValue.toFixed(4)}`);
  });

  it('applies the correct styles', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <HorizontalToken
            tokenIndex={tokenIndex}
            tokenLogo={logo}
            tokenSymbol={tokenSymbol}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
            isLast={isLast}
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;

    expect(tree.type).toBe('div');
    expect(tree.props.className).toContain('flex');
    expect(tree.props.className).toContain('w-full');
    expect(tree.props.className).toContain('justify-between');
    expect(tree.props.className).toContain('py-5');
    if (!isLast) {
      expect(tree.props.className).toContain('border-b');
      expect(tree.props.className).toContain('border-[#1F1D23]');
      expect(tree.props.className).toContain('mobile:border-[#27262F]');
    }
  });

  it('applies correct mobile border style for the last token when tokenIndex is 3', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <HorizontalToken
            tokenIndex={3}
            tokenLogo={logo}
            tokenSymbol={tokenSymbol}
            tokenName={tokenName}
            tokenValue={tokenValue}
            percentage={percentage}
            isLast
          />
        </Provider>
      )
      .toJSON() as ReactTestRendererJSON;

    expect(tree.type).toBe('div');
    expect(tree.props.className).toContain('mobile:border-b');
    expect(tree.props.className).toContain('mobile:border-[#27262F]');
  });
});
