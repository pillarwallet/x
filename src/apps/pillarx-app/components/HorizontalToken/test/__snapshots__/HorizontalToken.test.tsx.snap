// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HorizontalToken /> renders correctly and matches snapshot 1`] = `
<div
  className="flex w-full justify-between py-5 cursor-pointer border-b border-[#1F1D23] mobile:border-[#27262F] false"
>
  <div
    className="flex items-center"
  >
    <BodySmall
      className="text-purple_light mr-[18px]"
    >
      0
      1
    </BodySmall>
    <div
      className="relative w-[50px] h-[50px] rounded-full mr-3.5"
    >
      <img
        alt="token-logo"
        className="w-full h-full object-fill rounded-full"
        data-testid="horizontal-token-logo"
        src="https://example.com/logo.png"
      />
    </div>
    <div
      className="flex flex-col"
    >
      <Body>
        TNS
      </Body>
      <BodySmall
        className="text-purple_light"
      >
        Token Name
      </BodySmall>
    </div>
  </div>
  <div
    className="flex flex-col items-center"
  >
    <Body>
      $
      123.4567
    </Body>
    <TokensPercentage
      percentage={12.34}
    />
  </div>
</div>
`;
